---
title: "mark_two"
author: "patrick"
date: "30/04/2019"
output: html_document
---

### Clustering


```{r pack, echo=FALSE}
library(cluster)
library(factoextra)
library(gridExtra)
library(tidyverse)
```

## Clustering Setup
```{r setup}
setwd('/Users/patrickhedley-miller/code/R/infxRNAseq')
# setwd('/Users/patrickhedley-miller/code/gitWorkspace/infxRNAseq')

rm(list=setdiff(ls(), 'all'))
load('esets.RData')

rm(e.set.i, status.iris, targets, targets.iris)
# dim(e.set.i) # iris data
# dim(status.iris) # iris labels

# transpose
e.set.t <- t(e.set)

label <- as.character(status$most_general)

idx <- (label == 'bacterial' | label =='viral' |
          label == 'greyb' | label =='greyv')

e.set <- e.set.t[idx,]
label <- label[idx]
# rm(e.set.t)


### DGE
bct <- e.set[label == 'bacterial',]
vrl <- e.set[label == 'viral',]
dim(bct)
dim(vrl)

bct.mean <- apply(bct, 2, mean)
vrl.mean <- apply(vrl, 2, mean)


# Just get the maximum of all the means
limit = max(bct.mean, vrl.mean)

# Scatter plot
plot(vrl.mean ~ bct.mean, xlab = "bct", ylab = "vrl",
     main = "Bct vs Vrl - Scatter", xlim = c(3.5, limit), ylim = c(0, limit))
cor(bct.mean, vrl.mean)

# Compute fold-change (biological significance)
fold = bct.mean - vrl.mean

# Histogram of the fold differences
hist(fold, col = "gray")

# Compute statistical significance (using t-test)
pvalue = NULL # Empty list for the p-values
tstat = NULL # Empty list of the t test statistics

for(i in 1 : ncol(bct)) { # For each gene : 
  x = bct[,i] 
  y = vrl[,i]
  t = t.test(x, y)
  pvalue[i] = t$p.value
  tstat[i] = t$statistic
}

# Histogram of p-values (-log10)
hist(-log10(pvalue), col = "gray")

# Volcano: put the biological significance (fold-change)
# and statistical significance (p-value) in one plot
plot(fold, -log10(pvalue), main = "Volcano #1")

fold_cutoff = 1.5
pvalue_cutoff = 0.00000001
abline(v = fold_cutoff, col = "blue", lwd = 2)
abline(v = -fold_cutoff, col = "red", lwd = 2)
abline(h = -log10(pvalue_cutoff), col = "green", lwd = 2)


# Fold-change filter for "biological" significance
filter_by_fold = abs(fold) >= fold_cutoff
# dim(e.set[,filter_by_fold])

# P-value filter for "statistical" significance
filter_by_pvalue = pvalue <= pvalue_cutoff
# dim(e.set[,filter_by_pvalue])

# Combined filter (both biological and statistical)
filter_combined = filter_by_fold & filter_by_pvalue
# sum(filter_combined)

filtered = e.set[,filter_combined]
# dim(filtered)

# e.set[,filter_combined][1:10,1:5]
e.set.f <- e.set[,filter_combined]
# class(e.set.f)

```



```{r cluster_selection, echo=FALSE}
# clustering of DB, PB, DV, PV based on the 50 most differentially expressed genes identified above.
fviz_nbclust(e.set.f, kmeans, method = "wss")
fviz_nbclust(e.set.f, kmeans, method = "silhouette")
```

## cluster_two
```{r cluster_2, echo=FALSE}
## K2
k2 <- kmeans(e.set.f, centers = 2, nstart = 25)
str(k2)

k2$cluster <- as.factor(k2$cluster)

sig.1 <- as.numeric(e.set.f[,colnames(e.set.f)[1]])
sig.2 <- as.numeric(e.set.f[,colnames(e.set.f)[2]])

ggplot(as.data.frame(e.set.f), aes(sig.1, sig.2, color = k2$cluster, shape=label)) +
  geom_point(size=2) +
  xlab(paste('transcript',colnames(e.set.f))[1]) +
  ylab(paste('transcript',colnames(e.set.f))[2]) +
  ggtitle("Pairwise Scatter Plot of Most Significantly Differentially Expressed Transcripts")

table(k2$cluster, label)

e.set.pca <- prcomp(e.set.f, scale = TRUE)
# summary(e.set.pca)
plot(e.set.pca, type = 'l')

pair1 <- e.set.pca$x[,1:2]
pair2 <- e.set.pca$x[,3:4]

fviz_cluster(k2, geom = c("point"),  data = e.set.f, axes = c(1,2)) +
  ggtitle("PCA Cluster Assignment k = 2")

ggplot(as.data.frame(pair1), aes(PC1, PC2)) + geom_point(aes(color=k2$cluster, shape=label)) +
  xlab("First Principal Component") +
  ylab("Second Principal Component") +
  ggtitle("Cluster Assignment of First Two Principal Components")
```


## cluster_three
```{r cluster_3, echo=FALSE}
k3 <- kmeans(e.set.f, centers = 3, nstart = 25)
str(k3)
k3$cluster <- as.factor(k3$cluster)

ggplot(as.data.frame(e.set.f), aes(sig.1, sig.2, color = k3$cluster, shape=label)) +
  geom_point(size=2) +
  xlab(paste('transcript',colnames(e.set.f))[1]) +
  ylab(paste('transcript',colnames(e.set.f))[2]) +
  ggtitle("Pairwise Scatter Plot of Most Significantly Differentially Expressed Transcripts")

table(k3$cluster, label)

fviz_cluster(k3, geom = c("point"),  data = e.set.f, axes = c(1,2)) +
  ggtitle("PCA Cluster Assignment k = 3")

ggplot(as.data.frame(pair1), aes(PC1, PC2)) + geom_point(aes(color=k3$cluster, shape=label)) +
  xlab("First Principal Component") +
  ylab("Second Principal Component") +
  ggtitle("Cluster Assignment of First Two Principal Components")


```


## cluster_four
```{r cluster_4, echo=FALSE}
k4 <- kmeans(e.set.f, centers = 4, nstart = 25)
str(k4)
k4$cluster <- as.factor(k4$cluster)

ggplot(as.data.frame(e.set.f), aes(sig.1, sig.2, color = k4$cluster, shape=label)) +
  geom_point(size=2) +
  xlab(paste('transcript',colnames(e.set.f))[1]) +
  ylab(paste('transcript',colnames(e.set.f))[2]) +
  ggtitle("Pairwise Scatter Plot of Most Significantly Differentially Expressed Transcripts")

table(k4$cluster, label)

fviz_cluster(k4, geom = c("point"),  data = e.set.f, axes = c(1,2)) +
  ggtitle("PCA Cluster Assignment K = 4")

ggplot(as.data.frame(pair1), aes(PC1, PC2)) + geom_point(aes(color=k4$cluster, shape=label)) +
  xlab("First Principal Component") +
  ylab("Second Principal Component") +
  ggtitle("Cluster Assignment of First Two Principal Components")

```


