---
title: "mark_three"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(plyr)
library(limma)
library(cluster)
library(factoextra)
library(ggplot2)
require(reshape) # for melt()
require(scales) # for percent
library(gridExtra)
library(dplyr)
```



```{r block_1, echo=FALSE}
getwd()
setwd('/home/patrick/Code/R')
# setwd('/Users/patrickhedley-miller/code/gitWorkspace/infxRNAseq')

rm(list=setdiff(ls(), 'all'))
load('esets.RData')

e.set.t <- t(e.set)

idx <- status['most_general'] == 'bacterial' |
  status['most_general'] == 'viral' |
  status['most_general'] == 'greyb' |
  status['most_general'] == 'greyv'
  # status['most_general'] == 'OD'


clean<-union(which(as.character(status[idx,]$array.contemporary.CRP) == 'na'), which(status[idx,]$WBC == 0))
clean.idx <- seq(1:nrow(status[idx,]))[-(clean)]
```



```{r block_2, echo=FALSE}
# Diagnosis Breakdown
table(droplevels(status[idx,]$most_general))

dx.cols <- c("#ed0404", "#fc5716", '#16fc31', "#165bfc")
dx.cols.f <- c("#ed0404", "#fc5716",'#16fc31', '#16e1fc', '#165bfc', "#7a16fc", '#fc16f4')

positions <- c('bacterial', 'greyb', 'greyv', 'viral')
positions.f <- c('bacterial', 'greyb', 'greyv', 'flu', 'RSV', 'adeno', 'viralother')

ggplot(status[idx,], aes(most_general, fill=most_general, alpha=Sex)) +
  scale_alpha_manual(values=c(0.6, 1)) +
  labs(title = "Barplot of Diagnostic Group Breakdown by Gender", x = "Diagnosis", y = "Counts")+
  scale_x_discrete(limits = positions)+
  scale_fill_manual(values=dx.cols)+
  geom_bar()

ggplot(status[idx,], aes(most_general, group=Sex)) + 
  geom_bar(aes(y = ..prop.., fill=factor(..x..)), stat="count") +
  geom_text(aes( label = scales::percent(..prop..),
                 y= ..prop.. ), stat= "count", vjust = -.5) +
  scale_fill_manual(values=dx.cols)+
  facet_grid(~Sex) +
  coord_flip()+
  scale_y_continuous(labels = scales::percent)+
  labs(title = "Barplot of Percentage Diagnostic Group Breakdown by Gender", x = "Diagnosis", y = "Percentage")

table(droplevels(status[idx,]$more_general))
ggplot(status[idx,], aes(more_general, group=Sex)) + 
  geom_bar(aes(y = ..prop.., fill=factor(..x..)), stat="count") +
  geom_text(aes( label = scales::percent(..prop..),
                 y= ..prop.. ), stat= "count") +
  scale_fill_manual(values=dx.cols.f)+
  scale_x_discrete(limits=positions.f)+
  facet_grid(~Sex) +
  coord_flip()+
  scale_y_continuous(labels = scales::percent)+
  labs(title = "Barplot of Percentage Full Diagnostic Groups Breakdown by Gender", x = "Diagnosis", y = "Percentage")

table(status[idx,]$Sex)
chisq.test(table(status[idx,]$Sex))

table(droplevels(status[idx,]$most_general), status[idx,]$Sex)
# table(droplevels(status[idx,]$more_general), status[idx,]$Sex)

chisq.test(table(droplevels(status[idx,]$most_general), status[idx,]$Sex))
# chisq.test(table(droplevels(status[idx,]$more_general), status[idx,]$Sex))


# Age Breakdown
ggplot(status[idx,], aes(status[idx,]$most_general, Age..months., fill=Sex)) + geom_boxplot()+
  scale_x_discrete(limits = positions)+
  xlab('Diagnostic Group') +
  ylab('Age') +
  scale_x_discrete(limits = positions)+
  scale_fill_manual(values=c('#16acfc', '#fc1676'))+
  ggtitle("Box and Whisker Plot of Age by Diagnostic Group, Split by Gender")


ddply(status[idx,], ~most_general, summarise, mean=mean(Age..months.))

a <- as.data.frame(droplevels(status[idx,]$most_general))
colnames(a) <- 'dx'
a$age <-status[idx,]$Age..months.
b <- ddply(a,~dx)
anova.res <- aov(age ~ dx, data = b)
## ANOVA MEAN AGE BETWEEN GROUPS
summary(anova.res)


# Inflamatory Marker Breakdown
p1 <- ggplot(status[idx,][clean.idx,], aes(most_general, WBC, fill=Sex)) + geom_boxplot() +
  # scale_y_continuous(limits = c(0, 50))+
  ylab('WBC Count') +
  scale_x_discrete(limits = positions)+
  scale_fill_manual(values=c('#16acfc', '#fc1676'))+
  ggtitle("Box and Whisker Plot of WBC Count by Diagnosis, Split by Gender")

p2 <- ggplot(status[idx,][clean.idx,], aes(most_general,
  as.numeric(as.character(status[idx,][clean.idx,]$array.contemporary.CRP)), fill=Sex))+
  geom_boxplot()+
  # scale_y_continuous(limits = c(0, 230))+
  xlab('Cluster Assignment') +
  ylab('CRP Count') +
  scale_x_discrete(limits = positions)+
  scale_fill_manual(values=c('#16acfc', '#fc1676'))+
  ggtitle("Box and Whisker Plot of CRP Count by Diagnosis, Split by Gender")
gridExtra::grid.arrange(p1, p2, nrow = 2)

crp <- as.numeric(as.character(status[idx,][clean.idx,]$array.contemporary.CRP))
t.test(crp[status[idx,][clean.idx,]$Sex=='M' & status[idx,][clean.idx,]$most_general == 'bacterial'],
       crp[status[idx,][clean.idx,]$Sex=='F' & status[idx,][clean.idx,]$most_general == 'bacterial'])
# p-value = 0.6789
# no stat sig difference in crp between males and females in bct group



```


```{r block_3, echo=FALSE}
###### differential gene expression with limma ######

e.set.t <- t(e.set)
e.set.f <- e.set.t[idx,]
X <- as.matrix(t(e.set.f))
X.t <- t(X)

e.set.f <- as.data.frame(e.set.f)
e.set.f$label <- as.character(status[idx,c('most_general')])
e.set.f$sex <- status[idx,c('Sex')]
e.set.f$age <- status[idx,c('Age..months.')]

### DESIGN MATRIX
design <- model.matrix(~label + sex + age + 0, data = e.set.f)
colnames(design)<- c("bct","greyb","greyv", 'vrl', 'sexM', 'age')

design[1:10,]

contrast.matrix<- makeContrasts("bct-vrl", 'bct-greyv', levels=design)
contrast.matrix
# colnames(fit$coefficients)

fit <- lmFit(X, design)
fit2<- contrasts.fit(fit, contrast.matrix)
fit2 <- eBayes(fit2)

lfc <- 1.5
pval <- 0.05

results <- decideTests(fit2, method='global', p.value = pval, adjust.method = 'BH', lfc=lfc, coef = 'bct-vrl')
summary(results)
vennDiagram(results, include = 'both')

results.hits <- union(colnames(X.t)[results[,1] == 1],
                      colnames(X.t)[results[,1] == -1])

top.hits <- topTable(fit2, p.value = pval, adjust.method = 'BH', lfc=lfc, coef = 'bct-vrl')
all.hits <- topTable(fit2, number=nrow(fit2), coef = 'bct-vrl')

ggplot(all.hits, aes(y=-log10(adj.P.Val), x=logFC)) +
  geom_point(size=2) +
  geom_hline(yintercept = -log10(pval), linetype="longdash", colour="grey", size=1) +
  geom_vline(xintercept = lfc, linetype="longdash", colour="#BE684D", size=1) +
  geom_vline(xintercept = -(lfc), linetype="longdash", colour="#2C467A", size=1)+
  ggtitle("Volcano Plot of Log Fold Change Against -log10 P Value
    Cutoff - Fold Change:1, P Val:0.05")

```



```{r block_4, echo=FALSE}
### PCA
X.t <- t(X)
dim(X.t[,results.hits])
X.pca <- prcomp(X.t[,results.hits], scale = TRUE)

# summary(e.set.pca)
plot(X.pca, type = 'l')

pair1 <- as.data.frame(X.pca$x[,1:2])
pair2 <- as.data.frame(X.pca$x[,3:4])


fviz_nbclust(X.t[,results.hits], kmeans, method = "wss")
fviz_nbclust(X.t[,results.hits], kmeans, method = "silhouette")

gap_stat <- clusGap(X.t[,results.hits], FUN = kmeans, nstart = 25,
                    K.max = 20, B = 25)
fviz_gap_stat(gap_stat)

### K2
k2 <- kmeans(X.t[,results.hits], centers = 2, nstart = 500)
k2$cluster <- as.factor(k2$cluster)

k2.df <- status[idx,][clean.idx, c('my_category_2', 'most_general', 'more_general',
                                   'Age..months.', 'Sex', 'WBC', 'array.contemporary.CRP', 'Diagnosis')]
k2.df$cluster <- k2$cluster[clean.idx]
k2.df$array.contemporary.CRP <- as.numeric(as.character(k2.df$array.contemporary.CRP))

addmargins(table(k2$cluster, droplevels(status[idx,]$most_general)))
addmargins(table(k2$cluster, droplevels(status[idx,]$more_general)))

k2.clus.col <- c("#ed0404", "#165bfc")
positions.more <- c('bacterial', 'greyb', 'greyv', 'adeno', 'flu', 'RSV', 'viralother')
ggplot(status[idx,], aes(more_general, fill=k2$cluster)) +
  labs(title = "Barplot of Diagnostic Groups by Cluster", x = "Diagnosis", y = "Counts")+
  scale_x_discrete(limits = positions.more)+
  scale_fill_manual(values=k2.clus.col)+
  geom_bar()
# table(k2$cluster, status[idx, c('Sex')])

ggplot(pair1, aes(PC1, PC2)) + geom_point(aes(color=k2$cluster, shape=status[idx,]$most_general), size=2) +
  xlab("First Principal Component") +
  ylab("Second Principal Component") +
  scale_color_manual(values=k2.clus.col)+
  ggtitle("Cluster Assignment of First Two Principal Components")

ggplot(pair2, aes(PC3, PC4)) + geom_point(aes(color=k2$cluster, shape=status[idx,]$most_general), size=2) +
  xlab("Third Principal Component") +
  ylab("Fourth Principal Component") +
  scale_color_manual(values=k2.clus.col)+
  ggtitle("Cluster Assignment of Third-Fourth Principal Components")

ggplot(status[idx,], aes(most_general, Age..months., fill=k2$cluster)) + geom_boxplot()+
  scale_x_discrete(limits = positions)+
  xlab('Diagnostic Group') +
  ylab('Age') +
  scale_fill_manual(values=k2.clus.col)+
  ggtitle("Age (months) by Diagnostic Group, Split by Cluster")

p1.wbc <- ggplot(status[idx,][clean.idx,], aes(most_general, WBC, fill=k2$cluster[clean.idx])) + geom_boxplot() +
  # scale_y_continuous(limits = c(0, 50))+
  ylab('WBC Count') +
  xlab('') +
  scale_fill_manual(values=k2.clus.col)+
  ggtitle("Box Plot of WBC and CRP Count by Diagnosis")
p2.crp <- ggplot(status[idx,][clean.idx,], aes(most_general,
  as.numeric(as.character(status[idx,]$array.contemporary.CRP[clean.idx])), fill=k2$cluster[clean.idx]))+
  geom_boxplot()+
  ylab('CRP Count') +
  xlab('Diagnosis') +
  scale_fill_manual(values=k2.clus.col)
gridExtra::grid.arrange(p1.wbc, p2.crp, nrow = 2)


detach('package:plyr', unload = TRUE, character.only = TRUE)
k2.df %>%
  select(WBC, cluster, most_general, array.contemporary.CRP, Age..months., Sex) %>%
  group_by(cluster, most_general) %>%
  summarise(wbc.m = mean(WBC), crp.m = mean(array.contemporary.CRP), age.m = mean(Age..months.))

t.test(k2.df$WBC[k2.df$cluster == 1 & k2.df$most_general == 'bacterial'], k2.df$WBC[k2.df$cluster == 2 & k2.df$most_general == 'bacterial'])
# not enough power to detect sig diff within bct group given few bct assigned to class 2
# no difference with crp


```



```{r block_5, echo=FALSE}

### K3
k3 <- kmeans(X.t[,results.hits], centers = 3, nstart = 25)
# str(k3)
k3$cluster <- as.factor(k3$cluster)

addmargins(table(k3$cluster, droplevels(status[idx,]$most_general)))
addmargins(table(k3$cluster, droplevels(status[idx,]$more_general)))

k3.df <- status[idx,][clean.idx, c('my_category_2', 'most_general', 'more_general',
                                   'Age..months.', 'Sex', 'WBC', 'array.contemporary.CRP', 'Diagnosis')]
k3.df$cluster <- k3$cluster[clean.idx]
k3.df$array.contemporary.CRP <- as.numeric(as.character(k3.df$array.contemporary.CRP))

# had to un attach plyr to get working
k3.df %>%
  select(WBC, cluster, most_general, array.contemporary.CRP, Age..months., Sex) %>%
  group_by(cluster, Sex) %>%
  summarise(wbc.m = mean(WBC), crp.m = mean(array.contemporary.CRP), age.m = mean(Age..months.))

k3.clus.col <- c("#ed0404", "#fcac16", "#165bfc")
ggplot(status[idx,], aes(more_general, fill=k3$cluster)) +
  labs(title = "Barplot of Diagnostic Group Breakdown by Gender", x = "Diagnosis", y = "Counts")+
  scale_x_discrete(limits = positions.more)+
  scale_fill_manual(values=k3.clus.col)+
  geom_bar()

ggplot(pair1, aes(PC1, PC2)) + geom_point(aes(color=k3$cluster, shape=status[idx,]$most_general), size=2) +
  xlab("First Principal Component") +
  ylab("Second Principal Component") +
  scale_color_manual(values=k3.clus.col)+
  ggtitle("Cluster Assignment of First Two Principal Components")


k3.wbc <- ggplot(status[idx,][clean.idx,], aes(most_general, WBC, fill=k3$cluster[clean.idx])) + geom_boxplot() +
  # scale_y_continuous(limits = c(0, 50))+
  ylab('WBC Count') +
  scale_x_discrete(limits = positions)+
  scale_fill_manual(values=k3.clus.col)+
  ggtitle("Jitter Plot: WBC Count by Diagnosis")

k3.crp <- ggplot(status[idx,][clean.idx,], aes(most_general,
                                               as.numeric(as.character(status[idx,]$array.contemporary.CRP[clean.idx])), fill=k3$cluster[clean.idx]))+
  geom_boxplot()+
  # scale_y_continuous(limits = c(0, 230))+
  xlab('Cluster Assignment') +
  ylab('CRP Count') +
  scale_fill_manual(values=k3.clus.col)+
  ggtitle("Jitter Plot: CRP Count by Diagnosis")
gridExtra::grid.arrange(k3.wbc, k3.crp, nrow = 2)

k3.wbc <- ggplot(status[idx,][clean.idx,], aes(most_general, WBC, color=k3$cluster[clean.idx])) + geom_jitter(alpha = .9, width = 0.2) +
  # scale_y_continuous(limits = c(0, 50))+
  ylab('WBC Count') +
  scale_x_discrete(limits = positions)+
  scale_color_manual(values=k3.clus.col)+
  ggtitle("Jitter Plot: WBC Count by Diagnosis")

k3.crp <- ggplot(status[idx,][clean.idx,], aes(most_general,
                                               as.numeric(as.character(status[idx,]$array.contemporary.CRP[clean.idx])), color=k3$cluster[clean.idx]))+
  geom_jitter(alpha = .9, width = 0.2)+
  # scale_y_continuous(limits = c(0, 230))+
  xlab('Cluster Assignment') +
  ylab('CRP Count') +
  scale_color_manual(values=k3.clus.col)+
  ggtitle("Jitter Plot: CRP Count by Diagnosis")
gridExtra::grid.arrange(k3.wbc, k3.crp, nrow = 2)

```



```{r block_, echo=FALSE}

```

