<!DOCTYPE HTML>

<html>
	<head>
		<title>Massively by HTML5 UP</title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no" />
		<link rel="stylesheet" href="assets/css/main.css" />
		<!-- <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous"> -->
		<noscript><link rel="stylesheet" href="assets/css/noscript.css" /></noscript>
	</head>
	<body class="is-preload">

		<!-- Wrapper -->
			<div id="wrapper" class="fade-in">

				<!-- Intro -->
					<div id="intro">
						<h1>Clustering Microarray Data <br>for Class Discovery <br> in Febrile Children</h1>
						<ul class="actions">
							<li><a href="#header" class="button icon solo fa-arrow-down scrolly">Continue</a></li>
						</ul>
					</div>

				<!-- Header -->
					<header id="header">
						<a href="intro.html" class="logo">Introduction</a>
					</header>

				<!-- Nav -->
					<nav id="nav">
						<ul class="links">
							<li class='active'><a href="index.html">Home</a></li>
							<li><a href="intro.html">Intro</a></li>
							<li><a href="methods.html">Methods</a></li>
							<li><a href="results.html">Results</a></li>
							<li><a href="conclusion.html">Conclusion</a></li>
							<li><a href="acknowledgements.html">Acknowledgements</a></li>
							<li><a href="elements.html">elementss</a></li>
						</ul>
						<ul class="icons">
							<li><a href="#" class="icon fa-twitter"><span class="label">Twitter</span></a></li>
							<li><a href="#" class="icon fa-facebook"><span class="label">Facebook</span></a></li>
							<li><a href="#" class="icon fa-instagram"><span class="label">Instagram</span></a></li>
							<li><a href="#" class="icon fa-github"><span class="label">GitHub</span></a></li>
						</ul>
					</nav>

					<div id="main">
						<section> <!-- ABSTRACT LAY SUMMARY -->
							<header>
								<h2><a href="#">Lay Summary</a></h2>
							</header>
							<div class="box">
								<p>
									The discrimination of bacterial from viral infection in children still represents a major challenge for healthcare systems across the world. This distinction has important implications for treatment since delaying antibiotic therapy in bacterial infection can be life threatening whereas overuse of antibiotics can lead to the development of multi-drug resustant pathogens.
									Separating out these two diagnostic categories based on the patients symtpoms is difficult becuase children suffering from bacterial and viral infections will often have similar presentations.
									Recent <a href='https://jamanetwork.com/journals/jama/fullarticle/2545687'>work done by Professor Levin's group</a> at Imperial has demonstrated that gene expression levels measured during infection can distinguish with high accuracy bacterial from viral infection. This was achieved using a supervised machine learning model which was trained on the genes that were most different between the bacterial and viral cases.
									However bacterial and viral infections are both heterogeneous cohorts which are themselves composed of many different pathogens. The aim of this project has been to use a fully unsupervised methodology to group the patients based on the totality of there gene expression rather than genes that are most different between bacteria and viral infections.
									The hope is that by looking in an unbiased way at the entire gene expression profile we may start to unpick characteristic signatures that distinguish particular bacteria or particular viruses. This would be clinically useful because antibiotic therapy is most effective when targeted to a particular pathogen.
								</p>
							</div>
							<header>
								<h2><a href="#">Abstract</a></h2>
							</header>
							<div class="box">
									<dl>
										<dt>Background:</dt>
										<dd>
											<p>
												Transcriptomic signatures can be used to distinguish bacterial from viral infection with a high degree of accuracy. Clinicians can then use this information to guide theraputic decisions reducing the chance of missing bacterial infections and minimizing unnecessary use of antibiotics in patients with viral infections. However for the best treatment outcomes antibiotic therapy should be targeted to a particular pathogen.
											</p>
										</dd>
										<dt>Objective:</dt>
										<dd>
											<p>
												To use a fully unsupervised methodology to cluster patients based on the totality of their transcriptomic expression to identify subgroups which might have clinically relevant implications.
											</p>
										</dd>
										<dt>Methods:</dt>
										<dd>
											<p>
												I have used K-means clustering to optimally partition 35,959 transcripts from 239 febrile children and 62 health controls who presented to participating hospitals in the United Kingdom, Spain, the Netherlands, and the United States between 2009-2013.
												I have then used a supervised methodology to cluster the patients based on significantly differentially expressed transcripts between health and disease.
											</p>
										</dd>
										<dt>Results:</dt>
										<dd>
											<p>
												Cluster assignment of patents to 4 clusters shows a statistically significant relationship between the the cluster assignment and the causative pathogen.
												This is particularly evident with meningococcal disease with thirteen of seventeen cases assigned to cluster 2.
												These class assignments are robust and are preserved when using a supervised methodology.
											</p>
										</dd>
										<dt>Conclusion:</dt>
										<dd>
											<p>
												Fully unsupervised clustering of transcriptomic data from febrile children is able to identify clinically meaningful subgroups that are driven by the causative pathogen.
												The extent to which this clustering is determined by cell counts would require further de-convolution analysis.
											</p>
										</dd>																				
									</dl>
								<p>
								</p>
							</div>
							<header id="header">
									<a href="intro.html" class="logo">Introduction</a>
								</header>
						</section>
						<!-- Footer -->
							<!-- <footer>
								<div class="pagination">
									<a href="#" class="previous">Prev</a>
									<a href="#" class="page active">1</a>
									<a href="#" class="page">2</a>
									<a href="#" class="page">3</a>
									<span class="extra">&hellip;</span>
									<a href="#" class="page">8</a>
									<a href="#" class="page">9</a>
									<a href="#" class="page">10</a>
									<a href="#" class="next">Next</a>
								</div>
							</footer> -->
					</div>

					<div id="copyright">
						<ul><li>Patrick Hedley-Miller</li><li>Myrsini Kaforou</li></ul>
					</div>

			</div>

		<!-- Scripts -->
			<script src="assets/js/jquery.min.js"></script>
			<script src="assets/js/jquery.scrollex.min.js"></script>
			<script src="assets/js/jquery.scrolly.min.js"></script>
			<script src="assets/js/browser.min.js"></script>
			<script src="assets/js/breakpoints.min.js"></script>
			<script src="assets/js/util.js"></script>
			<script src="assets/js/main.js"></script>
			<!-- <script src="https://code.jquery.com/jquery-3.3.1.slim.min.js" integrity="sha384-q8i/X+965DzO0rT7abK41JStQIAqVgRVzpbzo5smXKp4YfRvH+8abtTE1Pi6jizo" crossorigin="anonymous"></script> -->
			<!-- <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js" integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1" crossorigin="anonymous"></script> -->
			<!-- <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script> -->
	</body>
</html>