<!DOCTYPE HTML>

<html>
	<head>
		<title>Generic Page - Massively by HTML5 UP</title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no" />
		<link rel="stylesheet" href="assets/css/main.css" />
		<noscript><link rel="stylesheet" href="assets/css/noscript.css" /></noscript>
	</head>
	<body class="is-preload">

		<!-- Wrapper -->
			<div id="wrapper">

				<!-- Header -->
					<header id="header">
						<a href="index.html" class="logo">Massively</a>
					</header>

				<!-- Nav -->
					<nav id="nav">
						<ul class="links">
							<li><a href="index.html">Home</a></li>
							<li><a href="intro.html">Intro</a></li>
							<li class="active"><a href="methods.html">Methods</a></li>
							<li><a href="results.html">Results</a></li>
							<li><a href="conclusion.html">Conclusion</a></li>
							<li><a href="acknowledgements.html">Acknowledgements</a></li>
							<!-- <li class="active"><a href="generic.html">Generic Page</a></li> -->

							<!-- <li><a href="elements.html">Elements Reference</a></li> -->
						</ul>
						<ul class="icons">
							<li><a href="#" class="icon fa-twitter"><span class="label">Twitter</span></a></li>
							<li><a href="#" class="icon fa-facebook"><span class="label">Facebook</span></a></li>
							<li><a href="#" class="icon fa-instagram"><span class="label">Instagram</span></a></li>
							<li><a href="#" class="icon fa-github"><span class="label">GitHub</span></a></li>
						</ul>
					</nav>

				<!-- Main -->
                <div id="main">
				<section class="post">
						<header class="major">
							<h1>Data Preparation</h1>
						</header>
						<p>
							I begin by filtering the microarray data based on mean expression. There is a risk when analysing microarray data that transcripts with low expression levels have a higher tendency to contribute noise rather than valid biological signal. Figure 4 below shows the relationship between log2 mean transcript expression on the X axis against log2 transcript variance on the y axis for the 47,323 genes in the dataset. The vertical line indicates a mean transcript expression of 5. The initial trend we see is that the transcript variance decreases as mean expression increases up until the threshold of 5.
							This indicates there is higher variability in the transcript results and potentially greater uncertanty, for genes that had lower expression levels. Excluding transcripts from down stream analysis whos mean expression was below the threshold reduces the chance that patients will be clustered based on erroneous transcripts whos variance was a product of low expression rather than biology.
							The cutoff of a mean expression of 5 was chosen based on domain knowledge and appears coherent with the trend seen in the Figure 4.
							Additionally excluding genes with lower mean expression will mean an increased ability to detect differentially expressed genes when using a Semi-Supervised approach becuase of the reduced number of tests that need to be adjusted for in multiple test correction.
							Following this procedure 35,970 genes remained for analysis.
						</p>
						<span class="image fit"><img src="images/mean_variance_scatter.png" alt="" /></span>
					</section>					
					<section class="post">
						<header class="major">
							<h1>Unsupervised<br>Clustering</h1>
						</header>
						<h2>K-Means Clustering</h2>
						<p>
							In the fully unsupervised stream I have used the K-means clustering algorithm to find the optimal cluster assignment for each patient based on the full 35,970 transcripts.
							Healthy control cases have been removed in the fully unsupervised stream since they were included in order to establish the list of differentially expressed genes between health and disease. As I am clustering based on all transcripts there is no need to select differentially expressed genes and hence I can remove the controls leaving 239 patients in the sample.
							I have chosen to use the Euclidian distance metric, defined in Figure XYZ below, to represent the (dis)similarity between pairs of observations x and y.
						</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/euclidian_distance.svg" alt="" /></span></div>
							</div>
						</div>

						<p>The basic intuition of the K-means algorithm is to define clusters so that the within cluster variation is minimized. The within cluster variation is defined as the sum of squared Euclidian distances between items and the centroid to which they have been assigned:</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/sum_squares.svg" alt="" /></span></div>
							</div>
						</div>
						<p>where:
							<ul>
								<li>x<sub>i</sub> is a data point belonging to the cluster C<sub>k</sub></li>
								<li>μ<sub>k</sub> is the mean value of the points assigned to the cluster C</li>
							</ul>
						</p>
						<p>Each observation x<sub>i</sub> is assigned to a given cluster so that the sum of squared distances between the observation and their assigned cluster centers μ<sub>k</sub> is minimized.</p>
						<p>By summing over all the K clusters we compute the total sum squares.</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/tot_sum_squares.svg" alt="" /></span></div>
							</div>
						</div>
						<h2>Optimal Number of Clusters</h2>
						<p>The K-means algorithm requires that the number of clusters be predetermined by the user. Three different statistical methods for selecting the optimal number of clusters are shown in Figure XYZ below. Figure .1 shows the total within sum of squares computed for incrementing numbers of clusters. Visual inspection of the graph can then be used to identify an inflection point beyond which increasing the number of clusters provides only a marginal decrease in the total sum of squares. Figure .2 shows the optimal number of clusters as computed by the Average Silhouette score. The Silhouette score determines how well a data point suited to its cluster with higher values indicating a better cluster cohesion. Taking the average over all data points gives the Average Silhouette score which is maximised at 2 clusters in this case. Figure .3 shows the optimal number of clusters as computed using the gap statistic as described by <a href='http://web.stanford.edu/~hastie/Papers/gap.pdf'>R. Tibshirani, G. Walther, and T. Hastie</a> in 2001. The gap statistic compares total within cluster variation for different values of k with their expected value under a null reference distribution. The null reference distirbution is generated by Monte Carlo simulation by generating B coppies of the reference dataset and computing the total within cluster variation. The gap statistic measres the deviation between the observed total within cluster variation and that expected under the null hypothesis. The optimal number of clusters is the value that maximises the gap statistic. The greatest reduction in variance occurs when the patients are partitoned into 2 clusters and this is also the optimal number as suggested using the average silhouette method. However the gap statistic suggests 4 clusters is optimal and this appears to be supported in Figure .1. Therefore we will examine the clinical data following clustering with k=2 and k=4.</p>
						<div class="box alt">
							<div class="row gtr-50 gtr-uniform">
								<div class="col-12">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz1992/64/#/.embed"
											height="500" width="100%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>								
								<div class="col-6">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz1992/66/#/.embed"
											height="400" width="100%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
								<div class="col-6">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz1992/68/#/.embed"
											height="400" width="100%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
							</div>
						</div>

					</section>

					<section class="post">
						<header class="major">
							<h1>Semi-Supervised<br>Clustering</h1>
						</header>
						A semi-supervised approach was then used in order to evaluate the robustness of the cluster assignments using the unsupervised approach.
						In the supervised stream the K-means algorithm was passed genes that were significantly different between health and disease.
						Clustering of the cases could then be compared to that produced by the unsupervised stream to check whether patients were assigned to the same or to different clusters.

						I used the <a href='https://academic.oup.com/nar/article/43/7/e47/2414268'>limma</a> bioconductor package to select differentially expressed genes between disease and health. The limma package allows the user to fit a linear model to the expresseion of each gene based on custom design matrix. The advantage of using this system rather than a simple pairwise t-test are manyfold.
						We saw in the clinical data analysis that diagnostic classes were not independent of patient gender and age. We can include these as co-variates in addition to the diagnostic class when constructing the design matrix and hence we adjust for them when selecting differentially expressed genes. Limma also allows us to construct a contrast matrix in which we define the comparisons we would like to make. This allows the selection disease agnostic differentially expressed genes as we can pass all 6 diagnostic groups into the matrix and set up a comparison between the healthy controls against a sumation of all of the other disease groups. 
						Finally, limma also uses Empirical Bayes to calculate global parameters across the entire gene set. The gene specific linear models can then be adjusted based on information that is shared across the entire gene set. The formula for the test statistic is shown in Figure XYZ below where t is the test statistic, &Beta;, &beta; is the linear model coefficient, s is the sample variance and n is the sample size. Samples with lower numbers of replicates will often fail to show significance becuase of the sample variance over the square root of the sample size will be large. The empirical bayes method adjusts the gene specific variance by the global variance computed across all genes This has the effect of increasing the effective degrees of freedom with which gene specific variances are estimated.

						
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/t_stat.svg" alt="" /></span></div>
							</div>
						</div>											
					</section>					
                </div>
                

				<!-- Footer -->
					<footer id="footer">
					
						<section class="split contact">
							<section>
								<h3>Social</h3>
								<ul class="icons alt">
									<!-- <li><a href="#" class="icon alt fa-twitter"><span class="label">Twitter</span></a></li>
									<li><a href="#" class="icon alt fa-facebook"><span class="label">Facebook</span></a></li>
									<li><a href="#" class="icon alt fa-instagram"><span class="label">Instagram</span></a></li> -->
									<li><a href="#" class="icon alt fa-github"><span class="label">GitHub</span></a></li>
								</ul>
							</section>
						</section>
					</footer>

				<!-- Copyright -->
					<div id="copyright">
						<ul><li>&copy; Untitled</li><li>Design: <a href="https://html5up.net">HTML5 UP</a></li></ul>
					</div>

			</div>

		<!-- Scripts -->
			<script src="assets/js/jquery.min.js"></script>
			<script src="assets/js/jquery.scrollex.min.js"></script>
			<script src="assets/js/jquery.scrolly.min.js"></script>
			<script src="assets/js/browser.min.js"></script>
			<script src="assets/js/breakpoints.min.js"></script>
			<script src="assets/js/util.js"></script>
			<script src="assets/js/main.js"></script>

	</body>
</html>