<!DOCTYPE HTML>

<html>
	<head>
		<title>Generic Page - Massively by HTML5 UP</title>
		<meta charset="utf-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no" />
		<link rel="stylesheet" href="assets/css/main.css" />
		<noscript><link rel="stylesheet" href="assets/css/noscript.css" /></noscript>
	</head>
	<body class="is-preload">

		<!-- Wrapper -->
			<div id="wrapper">

				<!-- Header -->
					<header id="header">
						<a href="index.html" class="logo">Methods</a>
					</header>

				<!-- Nav -->
					<nav id="nav">
						<ul class="links">
							<li><a href="index.html">Home</a></li>
							<li><a href="intro.html">Intro</a></li>
							<li class="active"><a href="methods.html">Methods</a></li>
							<li><a href="results.html">Results</a></li>
							<li><a href="discussion.html">Discussion</a></li>
							<li><a href="acknowledgements.html">Acknowledgements</a></li>
							<!-- <li class="active"><a href="generic.html">Generic Page</a></li> -->

							<!-- <li><a href="elements.html">Elements Reference</a></li> -->
						</ul>
						<ul class="icons">
							<li><a href="https://github.com/vetronz/RNA_seq_classifier" class="icon alt fa-github"><span class="label">GitHub</span></a></li>		
						</ul>
					</nav>

				<!-- Main -->
                <div id="main">
				<section class="post">
						<header class="major">
							<h1>Quality Control</h1>
						</header>
						<p>
							I began by filtering the microarray data based on mean expression. Transcripts with lower expression tend to have higher variance. We can see this trend in Figure 6 below which shows the mean expression against the mean variance with a log2 transformation. Following recommendations based on domain knowledge I have filtered out transcripts with a mean expression below 5, indicated by the vertical line. This results in 35,959 transcripts with a mean expression above this threshold which I will use to cluster the patients.
							An additional benefit of this filtering process is an increased statistical power to detect significantly differentially expressed genes when using the semi-supervised methodology because we are making fewer comparisons.
						</p>
						<span class="image fit"><img src="images/mean_variance_scatter.png" alt="" /></span>
						<p style="font-size: small;"><b>
								Figure 6 - Mean variance scatter plot shows initially that transcripts with a lower mean expression have a higher variance.
						</b></p>
					</section>					
					<section class="post">
						<header class="major">
							<h1>Unsupervised<br>Clustering</h1>
						</header>
						<h2>K-Means Clustering</h2>
						<p>
							In the fully unsupervised stream I have used the K-means clustering algorithm to cluster the patients based on the full 35,959 transcripts.
							Healthy control cases have been removed in the fully unsupervised stream since they were only included in order to establish genes which differed as between health and disease. As I am clustering based on all transcripts there is no need to select differentially expressed genes and hence I can remove the controls, leaving 239 patients in the sample.
							I have chosen to use the Euclidian distance metric, defined below, to represent the (dis)similarity between pairs of observations x and y.
						</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/euclidian_distance.svg" alt="" /></span></div>
							</div>
						</div>

						<p>The basic intuition of the K-means algorithm is to define clusters such that variation is minimized within a cluster. The within cluster variation is defined as the sum of squared Euclidian distances between items and the centroid to which they have been assigned:</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/sum_squares.svg" alt="" /></span></div>
							</div>
						</div>
						<p>where:
							<ul>
								<li>x<sub>i</sub> is a data point belonging to the cluster C<sub>k</sub></li>
								<li>μ<sub>k</sub> is the mean value of the points assigned to the cluster C</li>
							</ul>
						</p>
						<p>Each observation x<sub>i</sub> is assigned to a cluster in order that the sum of squared distances between the observation and their assigned centroid μ<sub>k</sub> is minimized.</p>
						<p>By summing over all the K clusters we compute the total sum squares.</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/tot_sum_squares.svg" alt="" /></span></div>
							</div>
						</div>
						<br>
						<h2>Optimal Number of Clusters</h2>
						<p>
							The K-means algorithm requires the number of clusters to be predetermined by the user. Three different statistical methods for selecting the optimal number of clusters are shown in Figure 7 below. Figure 7.1 shows the total within sum of squares computed for incrementing numbers of clusters. Visual inspection of the graph can then be used to identify an inflection point beyond which increasing the number of clusters provides only a marginal decrease in the total sum of squares. Figure 7.2 shows the optimal number of clusters as computed by the average silhouette score. The silhouette score determines how well a data point is suited to its cluster, with higher values indicating a better cluster cohesion. Taking the average over all data points gives the average silhouette score which is maximized at 2 clusters in this case. Figure 7.3 shows the optimal number of clusters as computed using the gap statistic as described by <a href='http://web.stanford.edu/~hastie/Papers/gap.pdf'>R. Tibshirani, G. Walther, and T. Hastie</a> in 2001. The gap statistic compares total within cluster variation for different values of k with their expected value under a null reference distribution. The null reference distribution is generated by Monte Carlo simulation by generating B coppies of the reference data set and computing the total within cluster variation. The gap statistic measures the deviation between the observed total within cluster variation and that expected under the null hypothesis. The optimal number of clusters is the value that maximizes the gap statistic, being 4 clusters in this case.
							The greatest reduction in variance occurs when the patients are partitioned into 2 clusters and this is also the optimal number as suggested using the average silhouette method. However the gap statistic suggests 4 clusters is optimal which is supported by the total sum of squares method. Therefore we will examine the clinical data following clustering with k=2 and k=4.
						</p>
						<div class="box alt">
							<div class="row gtr-50 gtr-uniform">
								<div class="col-12">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz1992/64/#/.embed"
											height="500" width="100%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>								
								<div class="col-6">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz1992/66/#/.embed"
											height="400" width="100%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
								<div class="col-6">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz1992/68/#/.embed"
											height="400" width="100%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
							</div>
						</div>
						<p style="font-size: small;"><b>
								Figure 7 - Shows optimal number of clusters as computed using total sum of squares(7.1), Average silhouette score (7.2) and Gap statistic (7.3).
						</b></p>

					</section>

					<section class="post">
						<header class="major">
							<h1>Semi-supervised Clustering</h1>
						</header>
						<p>
							In the semi-supervised methodology I have used the <a href='https://academic.oup.com/nar/article/43/7/e47/2414268'>limma</a> bioconductor package to select differentially expressed genes between disease and health. I have then reperformed the clustering using these transcripts rather than the full 35,595 transcripts.
						</p>
						<p>
							The limma package allows the user to fit a linear model to the expression of each gene based on a custom design matrix. The advantage of using this system rather than a simple pairwise t-test is many-fold.
							We saw in the clinical data analysis that diagnostic classes were not independent of patient gender and age. We can include these as covariates in addition to the diagnostic class when constructing the design matrix and hence we adjust for them when selecting differentially expressed genes. Limma also allows us to construct a contrast matrix in which we define the comparisons we would like to make. We can use the contrast matrix to define a disease cohort composed of the summation of our disease groups, namely the bacterial, probable bacterial, unknown, probable viral and viral cases. This allows the selection of genes that are differentially expressed between health and disease.
							
							Finally, limma also uses Empirical Bayes to calculate global parameters across the entire gene set. The gene specific linear models can then be adjusted based on information that is shared across the entire gene set. The formula for the test statistic is shown below where t is the test statistic, &beta; is the linear model coefficient, s is the sample variance and n is the sample size. Samples with lower numbers of replicates will often fail to show significance because the denominator will be large. The empirical Bayes method adjusts the gene specific variance by the global variance computed across all genes. This effectively increases the degrees of freedom with which gene specific variances are estimated.
						</p>
						<div class="box alt">
							<div class="row gtr-100 gtr-uniform">
								<div style="display: block; margin: 0 auto;"><span class="image fit"><img src="images/Equations/t_stat.svg" alt="" /></span></div>
							</div>
						</div>
						<h2>Assessment of Cluster Robustness using Bootstrap Re-sampling Methods</h2>
						<p>
							In order to evaluate the robustness of the cluster assignments found using the unsupervised stream I used 5 bootstrap samples with varying p-value and log fold change cut offs to select varying numbers of transcripts which I then used to cluster the patients. Table 2 shows the 5 bootstrap samples with the associated p-value and log fold change thresholds. The p-value and log fold change cut offs become more conservative as the number of bootstrap samples increases. This results in a gradual reduction in the number of transcripts that still meet the significance and fold change criteria.
						</p>
						<div class="table-wrapper">
							<table>
								<thead>
									<tr>
										<th>Bootstrap Sample</th>
										<th>P-Value</th>
										<th>Log Fold Change</th>
										<th>Number Significant Transcripts</th>
									</tr>
								</thead>
								<tbody>
									<tr>
										<td>Sample 1</td>
										<td>0.1</td>
										<td>0.25</td>
										<td>7,141</td>
									</tr>
									<tr>
										<td>Sample 2</td>
										<td>0.1</td>
										<td>0.375</td>
										<td>4,060</td>
									</tr>
									<tr>
										<td>Sample 3</td>
										<td>0.1</td>
										<td>0.5</td>
										<td>2,295</td>
									</tr>
									<tr>
										<td>Sample 4</td>
										<td>0.05</td>
										<td>0.75</td>
										<td>823</td>
									</tr>
									<tr>
										<td>Sample 5</td>
										<td>0.05</td>
										<td>1</td>
										<td>330</td>
									</tr>
								</tbody>
							</table>
						</div>
						<p style="font-size: small;"><b>
								Table 2 - Shows number of transcripts passed to the K-means algorithm for the 5 bootstrap samples with incrementing p-value and log fold change thresholds.
						</b></p>

						<p>Figure 8 below shows the volcano plots for the 5 bootstrap samples with log2 fold change on the X axis and -log10 adjusted P value on the Y axis. The significantly differentially expressed transcripts were then passed to the K-means algorithm. Clustering was then performed with K=4 for each of the bootstrap samples and then the proportion of cases assigned to the same cluster as the unsupervised stream was calculated.  
						</p>
						<div class="box alt">
							<div class="row gtr-50 gtr-uniform">
								<div class="col-6">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz/29/#/.embed"
											height="400" width="95%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>								
								<div class="col-6">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz/31/#/.embed"
											height="400" width="95%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
								<div class="col-4">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz/33/#/.embed"
											height="400" width="95%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
								<div class="col-4">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz/35/#/.embed"
											height="400" width="95%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>
								<div class="col-4">
									<span class="image fit">
										<iframe src="https://plot.ly/~vetronz/37/#/.embed"
											height="400" width="95%"
											scrolling="yes" seamless="seamless"
											frameBorder="0">
										</iframe>
									</span>
								</div>																
							</div>
						</div>
						<p style="font-size: small;"><b>
								Figure 8 - Volcano Plots show the significantly differentially expressed transcripts which were used to cluster the patients in each of the 5 bootstrap samples.
						</b></p>
						<a href="results.html" style='display: block;' class="button">Results</a>							
					</section>					
                </div>
                

				<!-- Footer -->
					<footer id="footer">
						<section class="split contact">
							<section>
								<h3>Social</h3>
								<ul class="icons alt">
									<li><a href="https://github.com/vetronz/RNA_seq_classifier" class="icon alt fa-github"><span class="label">GitHub</span></a></li>
								</ul>
							</section>
						</section>
					</footer>

				<!-- Copyright -->
					<div id="copyright">
						<ul><li>&copy; Untitled</li><li>Design: <a href="https://html5up.net">HTML5 UP</a></li></ul>
					</div>

			</div>

		<!-- Scripts -->
			<script src="assets/js/jquery.min.js"></script>
			<script src="assets/js/jquery.scrollex.min.js"></script>
			<script src="assets/js/jquery.scrolly.min.js"></script>
			<script src="assets/js/browser.min.js"></script>
			<script src="assets/js/breakpoints.min.js"></script>
			<script src="assets/js/util.js"></script>
			<script src="assets/js/main.js"></script>

	</body>
</html>